### PR TITLE
Add type hints

### PR DESCRIPTION
**Pull request by Tinche**
This is the best we can do, typing-wise, until Mypy has better support for the ParamSpec PEP (as far as I know).

Unfortunately, Mypy will allow the async decorators to be applied to non-async functions. Still, it's a big improvement since the type information of wrapped functions/coroutines won't be lost any more.

I also refactored `aio.time()`, it was creating a function on every call? Naughty naughty.

### CODE REVIEW

**Comment by hynek** [comment]
Are we relying on implicit Any here?

**Comment by Tinche** [comment]
> Are we relying on implicit Any here?

I don't think so, at least not for the decorator case. Can you clarify?

**Comment by hynek** [comment]
It's just that it's missing most type hints so I wonder about that.

**Comment by Tinche** [comment]
True, but:
* Can't be worse than now
* The decorator use cases are probably the most important ones, and those are typed now

There's also the issue of prometheus-client being untyped, but that's not our problem (fortunately or unfortunately)

**Comment by Tinche** [comment]
Looks like latest Mypy can handle ParamSpec, so I've changed the PR to use it. This will give an error if you're annotating an ordinary function, but wraps coroutines properly.

**Comment by hynek** [comment]
Hmmm I wondered if we can use it, but is there a way to overload such that it works with functions returning awaitable too? I have Futures somewhere in my somewhere. 🤔

**Comment by Tinche** [comment]
Have you ever known me to shirk from a challenge? ;)

Give me an example and I will do my best to make it type.

Also please approve tests ;)

**Comment by hynek** [comment]
Something like:

```python
@time
def f() -> async.Future:
    return asyncio.Future()
```

basically? It makes a bit more sense if its part of something bigger. 🤔

**Comment by Tinche** [comment]
`asyncio.Future` is a subtype of `Awaitable` though, so it should typecheck.

```
import asyncio

from prometheus_async.aio import time


@time("test")
def f(i: int) -> asyncio.Future:
    return asyncio.Future()


reveal_type(f)
```
```
a01.py:11: note: Revealed type is "def (i: builtins.int) -> asyncio.futures.Future*[Any]"
```

Looking at the test failures, looks like we need `typing_extensions` on all versions. Is that ok?

**Comment by hynek** [comment]
Could you add a `typing_examples.py` that shows all of this off?

typing_extensions is fine.

**Comment by Tinche** [comment]
@hynek I added a file with examples, let me know if that's what you had in mind.

File: CHANGELOG.rst [hunk 1]
```diff
@@ -32,11 +32,12 @@ Deprecations:
 Changes:
 ^^^^^^^^
 
+
+- Add support for static typing when using ``time``, ``count_exceptions`` and ``track_inprogress`` as decorators.
```
Comment (hynek) [inline CHANGELOG.rst:36]:
```suggestion
- Add support for static typing when using ``prometheus_async.aio.time()``, ``prometheus_async.aio.count_exceptions()`` and ``prometheus_async.aio.track_inprogress()`` as decorators.
```

File: CHANGELOG.rst [hunk 2]
```diff
@@ -32,11 +32,12 @@ Deprecations:
...
 - Added support for `OpenMetrics <https://openmetrics.io>`_ exposition in ``prometheus_async.aio.web.server_stats()`` and thus ``prometheus_async.aio.web.start_http_server_in_thread()``.
   `#23 <https://github.com/hynek/prometheus-async/issues/23>`_
 
 
-
```
Comment (hynek) [inline CHANGELOG.rst:39]: pls remove this

File: pyproject.toml [hunk 3]
```diff
@@ -11,7 +11,7 @@ authors = [{name = "Hynek Schlawack", email = "hs@ox.cx"}]
 dynamic = ["version", "description"]
 requires-python = ">=3.7"
 # Keep in-sync with testenv:oldest!
-dependencies = ["prometheus_client >= 0.8.0", "wrapt"]
+dependencies = ["prometheus_client >= 0.8.0", "wrapt", "typing_extensions"]
```
Comment (hynek) [inline pyproject.toml:14]: If I read https://github.com/python/typing/commit/4ba98e8c348cd0b2e9b8cc67e6b99eb19a0ed433#diff-6e5e49effcec8f8c76f9823331948755c9cf6925872981dbb44a3c4f95f0b3f3 correctly, we need a at least 3.10.0.

```suggestion
dependencies = ["prometheus_client >= 0.8.0", "wrapt", "typing_extensions >= 3.10.0"]
```

File: src/prometheus_async/aio/_decorators.py [hunk 4]
```diff
@@ -15,12 +15,24 @@
 """
 Decorators for asyncio.
 """
-
```
Comment (hynek) [inline src/prometheus_async/aio/_decorators.py:18]: pls remove this. imports glued to comments look gross ;)

File: typing_examples.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
+"""Some examples of prometheus-async typing integration."""
+from asyncio import Future
```
Comment (hynek) [inline typing_examples.py:2]:
```suggestion
"""
Some examples of prometheus-async typing integration.
"""

from asyncio import Future
```

File: typing_examples.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+@time(REQ_DURATION)
+async def func(i: int) -> str:
+    return str(i)
+
+
+# The type of `func` is correct:
+# "def (i: builtins.int) -> typing.Coroutine*[Any, Any, builtins.str]"
+# reveal_type(func)
+
```
Comment (hynek) [inline typing_examples.py:20]: Can't you do the `# type: ignore` trick here?

**Review by hynek — CHANGES_REQUESTED** [review]
had a quick look, we need to talk about the overloads too tho :)

File: pyproject.toml (hunk 3 above)
Comment (Tinche) [reply pyproject.toml:14]: Not sure what do you mean, we need `typing_extensions` always? It works on 3.9, I'm using it.

File: CHANGELOG.rst (hunk 2 above)
Comment (Tinche) [reply CHANGELOG.rst:39]: What am I removing here? The empty line?

File: typing_examples.py (hunk 6 above)
Comment (Tinche) [reply typing_examples.py:23]: Done, and added the Mypy option.

File: CHANGELOG.rst (hunk 2 above)
Comment (hynek) [reply CHANGELOG.rst:39]: no…you're removing the absence of an empty line :D

File: pyproject.toml (hunk 3 above)
Comment (hynek) [reply pyproject.toml:14]: this is not a condition, this is the lowest version of typing_extensions that contains ParamSpec :D

**Review by hynek — CHANGES_REQUESTED** [review]

File: src/prometheus_async/aio/_decorators.py [hunk 7]
```diff
@@ -17,10 +17,23 @@
...
 from wrapt import decorator
 
 
+P = ParamSpec("P")
+R = TypeVar("R", bound=Awaitable)
+
+
+@overload
+def time(
+    metric,
```
Comment (hynek) [inline src/prometheus_async/aio/_decorators.py:34]: How about we define ourselves Protocols for those metrics defining exactly what we need in it? Guess we could import the classes from prometheus_client but i suspect that's potential danger zone.

File: pyproject.toml (hunk 3 above)
Comment (Tinche) [reply pyproject.toml:14]: Ahh I see ;)

File: src/prometheus_async/aio/_decorators.py (hunk 7 above)
Comment (Tinche) [reply src/prometheus_async/aio/_decorators.py:34]: Hm maybe.

**Comment by Tinche** [comment]
OK, I added multiple overloads everywhere. I guessed for Twisted, could you double-check for me?

File: src/prometheus_async/aio/_decorators.py [hunk 8]
```diff
@@ -16,11 +16,31 @@
...
+) -> Callable[[Callable[P, R]], Callable[P, R]]:
+    ...
+
+
+@overload
+def time(metric, future: Future[T]) -> Awaitable[T]:
```
Comment (hynek) [inline src/prometheus_async/aio/_decorators.py:40]: I think this is what's breaking CI. Future isn't subscriptable…I think I've run into that problem myself once.

**Comment by hynek** [comment]
I also have code similar to this:

```python
await time(
    SomeHistogram,,
    self._send_cmd(cmd, cmd.names),
)
```

with

```python
    async def _send_cmd(
        self, cmd: IDomainCommand, domains: list[str]
    ) -> tuple[list[str], IEPPResult]:
        ...
```

that gives me:

```
src/domains_svc/proxy_client.py:159: error: No overload variant of "time" matches argument types "Any", "Coroutine[Any, Any, Tuple[List[str], IEPPResult]]"
src/domains_svc/proxy_client.py:159: note: Possible overload variants:
src/domains_svc/proxy_client.py:159: note:     def time(metric: Any) -> Callable[[Callable[P, R]], Callable[P, R]]
src/domains_svc/proxy_client.py:159: note:     def [T] time(metric: Any, future: Future[T]) -> Awaitable[T]
```

I guess we need to add `typing.Coroutine` overloads?

**Comment by Tinche** [comment]
> I also have code similar to this:
> 
> ```python
> await time(
>     SomeHistogram,,
>     self._send_cmd(cmd, cmd.names),
> )
> ```
> 
> with
> 
> ```python
>     async def _send_cmd(
>         self, cmd: IDomainCommand, domains: list[str]
>     ) -> tuple[list[str], IEPPResult]:
>         ...
> ```
> 
> that gives me:
> 
> ```
> src/domains_svc/proxy_client.py:159: error: No overload variant of "time" matches argument types "Any", "Coroutine[Any, Any, Tuple[List[str], IEPPResult]]"
> src/domains_svc/proxy_client.py:159: note: Possible overload variants:
> src/domains_svc/proxy_client.py:159: note:     def time(metric: Any) -> Callable[[Callable[P, R]], Callable[P, R]]
> src/domains_svc/proxy_client.py:159: note:     def [T] time(metric: Any, future: Future[T]) -> Awaitable[T]
> ```
> 
> I guess we need to add `typing.Coroutine` overloads?

Oh interesting. We can relax the type constraint to be Awaitable. Doing it

**Review by hynek — APPROVED** [review]

**Comment by hynek** [comment]
Hopefully last snag:  our types don't allow for decorator-style usage for Twisted.

**Pull request merged**